File: archive/c/muse/data_structures/SinglyLinkedListTest.c
#include <muse/data_structures/SinglyLinkedList.h>
#include <muse/util/TestRunner.h>

bool testSinglyLinkedList(void) {
  size_t size = 8192;

  SinglyLinkedList *list = SinglyLinkedList_new();

  for (size_t i = 1; i <= size; i++) {
    SinglyLinkedList_append(list, (int)i);
  }

  if (SinglyLinkedList_size(list) != size) {
    return false;
  }

  for (size_t i = 0; i < size; i++) {
    if ((size_t)SinglyLinkedList_get(list, i) != i + 1) {
      return false;
    }
  }

  for (size_t i = 0; i < size; i++) {
    SinglyLinkedList_set(list, i, size - i);
  }

  for (size_t i = 0; i < size; i++) {
    if ((size_t)SinglyLinkedList_get(list, i) != size - i) {
      return false;
    }
  }

  for (size_t i = 0, j = size - 1; i < j; i++, j--) {
    int x = SinglyLinkedList_get(list, i);
    int y = SinglyLinkedList_get(list, j);

    SinglyLinkedList_remove(list, i);
    SinglyLinkedList_insert(list, i, y);
    SinglyLinkedList_remove(list, j);
    SinglyLinkedList_insert(list, j, x);
  }

  for (size_t i = 0; i < size; i++) {
    if ((size_t)SinglyLinkedList_get(list, i) != i + 1) {
      return false;
    }
  }

  for (size_t i = size; i >= 1; i--) {
    if ((size_t)SinglyLinkedList_back(list) != i) {
      return false;
    }
    SinglyLinkedList_eject(list);
  }

  return SinglyLinkedList_isEmpty(list);
}

int main(void) { TestRunner_pick(&testSinglyLinkedList); }

File: archive/java/muse/data_structures/SinglyLinkedListTest.java
import muse.data_structures.SinglyLinkedList;
import muse.util.TestRunner;

public class SinglyLinkedListTest {

  public static boolean testSinglyLinkedList() {
    int size = 8192;

    SinglyLinkedList<Integer> list = new SinglyLinkedList<>();

    for (int i = 1; i <= size; i++) {
      list.append(i);
    }

    if (list.size() != size) {
      return false;
    }

    for (int i = 0; i < size; i++) {
      if (list.get(i) != i + 1) {
        return false;
      }
    }

    for (int i = 0; i < size; i++) {
      list.set(i, size - i);
    }

    for (int i = 0; i < size; i++) {
      if (list.get(i) != size - i) {
        return false;
      }
    }

    for (int i = 0, j = size - 1; i < j; i++, j--) {
      int x = list.get(i);
      int y = list.get(j);

      list.remove(i);
      list.insert(i, y);
      list.remove(j);
      list.insert(j, x);
    }

    for (int i = 0; i < size; i++) {
      if (list.get(i) != i + 1) {
        return false;
      }
    }

    for (int i = size; i >= 1; i--) {
      if (list.back() != i) {
        return false;
      }
      list.eject();
    }

    return list.isEmpty();
  }

  public static void main(String[] args) {
    TestRunner.pick(SinglyLinkedListTest::testSinglyLinkedList);
  }
}
